File: SMB_PERSISTANCE+HASHES/Python/samba_setup.py
#!/usr/bin/env python3
import os
import sys
import pwd
import grp
import subprocess
import logging
import shutil
from typing import List, Dict, TypedDict


# TODO: REFACTORING THE CODE SO IT'S MORE READABLE AND MAINTAINABLE
# TODO: ADD ERROR HANDLING AND LOGGING
# TODO: SILENTLY HANDLE ERRORS WHERE APPROPRIATE
# TODO: SILENT MODE OPTION


# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(levelname)s - %(message)s'
)
logger = logging.getLogger(__name__)


class HashData(TypedDict):
    LM: str
    NT: str


def get_smb_hashes() -> Dict[str, HashData]:
    """Retrieve SMB password hashes from the system."""
    try:
        # Check if running as root
        if os.geteuid() != 0:
            logger.error("Need root privileges to retrieve hashes")
            return {}

        hashes: Dict[str, HashData] = {}

        # Use pdbedit to get user list and hashes
        try:
            output = subprocess.check_output(
                ['pdbedit', '-L', '-w'],
                stderr=subprocess.PIPE
            ).decode()
            
            for line in output.splitlines():
                if ':' in line:
                    fields = line.split(':')
                    if len(fields) >= 4:
                        username = fields[0]
                        lm_hash = fields[2]
                        nt_hash = fields[3]
                        hashes[username] = {
                            'LM': lm_hash,
                            'NT': nt_hash
                        }
            logger.info(f"Successfully retrieved {len(hashes)} SMB hashes")
            return hashes
        except subprocess.CalledProcessError as e:
            logger.error(f"Failed to retrieve hashes using pdbedit: {e}")
            return {}

    except Exception as e:
        logger.error(f"Error retrieving SMB hashes: {e}")
        return {}


def check_root_privileges() -> bool:
    """Check if script is running with root privileges."""
    return os.geteuid() == 0


def install_samba() -> bool:
    """Install Samba and create backup of original configuration."""
    try:
        logger.info("Updating package lists...")
        subprocess.run(['apt-get', 'update'], check=True)
        
        logger.info("Installing Samba...")
        subprocess.run(['apt-get', 'install', '-y', 'samba'], check=True)
        
        # Backup original configuration
        shutil.copy('/etc/samba/smb.conf', '/etc/samba/smb.conf.bak')
        logger.info("Created backup of original Samba configuration")
        return True
    except subprocess.CalledProcessError as e:
        logger.error(f"Failed to install Samba: {e}")
        return False


def get_share_details() -> dict:
    """Get share configuration details from user input."""
    config = {}
    config['sharename'] = input("Enter name for the new share: ")
    config['sharepath'] = input("Enter full path to share directory: ")
    config['browseable'] = input(
        "Make share browseable? (yes/no) [no]: "
    ).lower() or 'no'
    config['writable'] = input(
        "Make share writable? (yes/no) [no]: "
    ).lower() or 'no'
    config['guestok'] = input(
        "Allow guest access? (yes/no) [no]: "
    ).lower() or 'no'
    return config


def setup_share_directory(path: str) -> bool:
    """Create and set up the share directory with proper permissions."""
    try:
        if not os.path.exists(path):
            create = input(
                "Path doesn't exist. Create it? (yes/no) [yes]: "
            ).lower() or 'yes'
            if create == 'yes':
                os.makedirs(path, exist_ok=True)
                logger.info(f"Created directory: {path}")
            else:
                logger.error("Aborting setup - directory not found")
                return False
        return True
    except Exception as e:
        logger.error(f"Failed to create directory: {e}")
        return False


def configure_samba_users() -> List[str]:
    """Configure Samba users and return list of configured usernames."""
    users = []
    while True:
        username = input(
            "Enter username for Samba access (leave empty to finish): "
        )
        if not username:
            break

        try:
            # Check if user exists
            subprocess.run(['id', username], check=True)
        except subprocess.CalledProcessError:
            create_user = input(
                "User doesn't exist. Create new system user? (yes/no) [no]: "
            ).lower() or 'no'
            if create_user == 'yes':
                try:
                    subprocess.run(
                        ['useradd', '-m', '-s', '/usr/sbin/nologin', username],
                        check=True
                    )
                    subprocess.run(['passwd', '-l', username], check=True)
                except subprocess.CalledProcessError as e:
                    logger.error(f"Failed to create system user: {e}")
                    continue

        try:
            # Add to Samba
            subprocess.run(['smbpasswd', '-L', '-a', username], check=True)
            subprocess.run(['smbpasswd', '-L', '-e', username], check=True)
            users.append(username)
            logger.info(f"Successfully added Samba user: {username}")
        except subprocess.CalledProcessError as e:
            logger.error(f"Failed to configure Samba user: {e}")

    return users


def create_samba_config(config: dict, users: List[str]) -> bool:
    """Create Samba configuration file with secure settings."""
    try:
        with open('/etc/samba/smb.conf', 'w') as f:
            f.write("""[global]
   workgroup = WORKGROUP
   server role = standalone server
   security = user
   map to guest = bad user
   smb encrypt = required
   min protocol = SMB2_02
   unix password sync = yes
   log file = /var/log/samba/log.%m
   max log size = 1000
   logging = file
   panic action = /usr/share/samba/panic-action %d
   server multi channel support = yes

""")

            # Write share configuration
            f.write(f"""[{config['sharename']}]
   path = {config['sharepath']}
   browseable = {config['browseable']}
   writable = {config['writable']}
   guest ok = {config['guestok']}
""")

            # Add user restrictions if needed
            if config['guestok'] == 'no' and users:
                f.write(f"   valid users = {' '.join(users)}\n")

        logger.info("Created Samba configuration file")
        return True
    except Exception as e:
        logger.error(f"Failed to create Samba configuration: {e}")
        return False


def set_permissions(config: dict, users: List[str]) -> None:
    """Set appropriate permissions on share directory."""
    try:
        if config['guestok'] == 'no' and users:
            os.chown(
                config['sharepath'],
                pwd.getpwnam(users[0]).pw_uid,
                grp.getgrnam(users[0]).gr_gid
            )
            # SGID to maintain group ownership
            os.chmod(config['sharepath'], 0o2770)
        else:
            os.chmod(config['sharepath'], 0o2777)
            logger.warning("Share is world-writable!")
    except Exception as e:
        logger.error(f"Failed to set permissions: {e}")


def ensure_persistence() -> bool:
    """Ensure Samba service persistence through systemd."""
    try:
        subprocess.run(['systemctl', 'enable', 'smbd'], check=True)
        subprocess.run(['systemctl', 'restart', 'smbd'], check=True)
        logger.info("Enabled and started Samba service")
        return True
    except subprocess.CalledProcessError as e:
        logger.error(f"Failed to configure service persistence: {e}")
        return False


def verify_configuration() -> bool:
    """Verify Samba configuration."""
    try:
        subprocess.run(['testparm', '-s'], check=True)
        logger.info("Samba configuration verified successfully")
        return True
    except subprocess.CalledProcessError as e:
        logger.error(f"Configuration verification failed: {e}")
        return False


def display_share_info(share_name: str) -> None:
    """Display share access information."""
    try:
        host_ip = subprocess.check_output(
            ['hostname', '-I']
        ).decode().split()[0]
        hostname = subprocess.check_output(['hostname']).decode().strip()
        hostname = hostname.split('.')[0]
        print("\nSamba setup complete!")
        print("Share accessible at:")
        print(f"  Host: {host_ip}")
        print(f"  Share: \\\\{hostname}\\{share_name}")
    except Exception as e:
        logger.error(f"Failed to display share information: {e}")


def main() -> None:
    """Main function to orchestrate Samba setup."""
    print("[+] Welcome to the Samba setup script!")
    
    if not check_root_privileges():
        logger.error("Please run this script as root")
        sys.exit(1)

    if not install_samba():
        print("[-] Failed to install Samba. Exiting.")
        sys.exit(1)

    config = get_share_details()
    
    if not setup_share_directory(config['sharepath']):
        sys.exit(1)

    users = configure_samba_users()
    
    if not create_samba_config(config, users):
        sys.exit(1)
        
    set_permissions(config, users)
    
    if not ensure_persistence():
        sys.exit(1)
        
    if not verify_configuration():
        sys.exit(1)
        
    display_share_info(config['sharename'])

    # Retrieve and display SMB hashes
    print("\n[+] Retrieving SMB password hashes...")
    hashes = get_smb_hashes()
    if hashes:
        print("\nSMB Password Hashes:")
        for username, hash_data in hashes.items():
            print(f"User: {username}")
            print(f"LM Hash: {hash_data['LM']}")
            print(f"NT Hash: {hash_data['NT']}")
            print("-" * 50)
    else:
        print("[-] No SMB hashes retrieved")


if __name__ == "__main__":
    main()

File: SMB_PERSISTANCE+HASHES/C/samba_setup.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#define MAX_INPUT 256
#define MAX_USERS 10
#define MAX_HASH_LENGTH 64


/*
Build the program: make
Clean build files: make clean
Install system-wide: sudo make install
Uninstall: sudo make uninstall
*/




// Function to retrieve SMB hashes
void get_smb_hashes() {
    printf("\n[+] Retrieving SMB password hashes...\n");
    
    // Check if running as root
    if(geteuid() != 0) {
        fprintf(stderr, "Need root privileges to retrieve hashes\n");
        return;
    }

    // Use pdbedit to get user list and hashes
    FILE *fp = popen("pdbedit -L -w 2>/dev/null", "r");
    if(!fp) {
        fprintf(stderr, "Failed to execute pdbedit command\n");
        return;
    }

    char line[MAX_INPUT];
    int found = 0;
    printf("\nSMB Password Hashes:\n");
    
    while(fgets(line, sizeof(line), fp)) {
        char *username = strtok(line, ":");
        if(!username) continue;
        
        // Skip three fields to get to the hashes
        char *lm_hash = NULL;
        char *nt_hash = NULL;
        
        for(int i = 0; i < 3; i++) {
            char *field = strtok(NULL, ":");
            if(!field) break;
            if(i == 1) lm_hash = field;
            if(i == 2) nt_hash = field;
        }
        
        if(username && lm_hash && nt_hash) {
            printf("User: %s\n", username);
            printf("LM Hash: %s\n", lm_hash);
            printf("NT Hash: %s\n", nt_hash);
            printf("-------------------------------------------------\n");
            found = 1;
        }
    }
    
    if(!found) {
        printf("[-] No SMB hashes retrieved\n");
    }
    
    pclose(fp);
}

void run_command(const char *cmd) {
    printf("Executing: %s\n", cmd);
    if(system(cmd) != 0) {
        perror("Command failed");
        exit(EXIT_FAILURE);
    }
}

// Function to retrieve SMB hashes
void get_smb_hashes() {
    printf("\n[+] Retrieving SMB password hashes...\n");
    
    // Check if running as root
    if(geteuid() != 0) {
        fprintf(stderr, "Need root privileges to retrieve hashes\n");
        return;
    }

    // Use pdbedit to get user list and hashes
    FILE *fp = popen("pdbedit -L -w 2>/dev/null", "r");
    if(!fp) {
        fprintf(stderr, "Failed to execute pdbedit command\n");
        return;
    }

    char line[MAX_INPUT];
    int found = 0;
    printf("\nSMB Password Hashes:\n");
    
    while(fgets(line, sizeof(line), fp)) {
        char *username = strtok(line, ":");
        if(!username) continue;
        
        // Skip three fields to get to the hashes
        char *lm_hash = NULL;
        char *nt_hash = NULL;
        
        for(int i = 0; i < 3; i++) {
            char *field = strtok(NULL, ":");
            if(!field) break;
            if(i == 1) lm_hash = field;
            if(i == 2) nt_hash = field;
        }
        
        if(username && lm_hash && nt_hash) {
            printf("User: %s\n", username);
            printf("LM Hash: %s\n", lm_hash);
            printf("NT Hash: %s\n", nt_hash);
            printf("-------------------------------------------------\n");
            found = 1;
        }
    }
    
    if(!found) {
        printf("[-] No SMB hashes retrieved\n");
    }
    
    pclose(fp);
}

void run_command(const char *cmd) {
    printf("Executing: %s\n", cmd);
    if(system(cmd) != 0) {
        perror("Command failed");
        exit(EXIT_FAILURE);
    }
}

int main() {
    if(geteuid() != 0) {
        fprintf(stderr, "Must be run as root\n");
        exit(EXIT_FAILURE);
    }

    // Install dependencies
    run_command("apt-get update");
    run_command("apt-get install -y samba");
    run_command("cp /etc/samba/smb.conf /etc/samba/smb.conf.bak");

    char share_name[MAX_INPUT];
    char share_path[MAX_INPUT];
    char browseable[MAX_INPUT] = "no";
    char writable[MAX_INPUT] = "no";
    char guest_ok[MAX_INPUT] = "no";
    
    printf("Enter share name: ");
    fgets(share_name, MAX_INPUT, stdin);
    share_name[strcspn(share_name, "\n")] = 0;
    
    printf("Enter full path to share directory: ");
    fgets(share_path, MAX_INPUT, stdin);
    share_path[strcspn(share_path, "\n")] = 0;
    
    // Create directory
    if(access(share_path, F_OK) != 0) {
        char create[MAX_INPUT];
        printf("Path doesn't exist. Create? (yes/no) [yes]: ");
        fgets(create, MAX_INPUT, stdin);
        create[strcspn(create, "\n")] = 0;
        if(strlen(create) == 0 || strcasecmp(create, "yes") == 0) {
            char cmd[MAX_INPUT * 2];
            snprintf(cmd, sizeof(cmd), "mkdir -p %s", share_path);
            run_command(cmd);
        }
    }
    
    // User management
    char users[MAX_USERS][MAX_INPUT];
    int user_count = 0;
    
    if(strcasecmp(guest_ok, "no") == 0) {
        while(user_count < MAX_USERS) {
            char user[MAX_INPUT];
            printf("Enter username (empty to finish): ");
            fgets(user, MAX_INPUT, stdin);
            user[strcspn(user, "\n")] = 0;
            
            if(strlen(user) == 0) break;
            
            // Check if user exists
            char cmd[MAX_INPUT * 2];
            snprintf(cmd, sizeof(cmd), "id %s >/dev/null 2>&1", user);
            if(system(cmd) != 0) {
                char create[MAX_INPUT];
                printf("User doesn't exist. Create? (yes/no) [no]: ");
                fgets(create, MAX_INPUT, stdin);
                create[strcspn(create, "\n")] = 0;
                
                if(strcasecmp(create, "yes") == 0) {
                    snprintf(cmd, sizeof(cmd), "useradd -m -s /usr/sbin/nologin %s", user);
                    run_command(cmd);
                    snprintf(cmd, sizeof(cmd), "passwd -l %s", user);
                    run_command(cmd);
                    snprintf(cmd, sizeof(cmd), "smbpasswd -L -a %s", user);
                    run_command(cmd);
                    snprintf(cmd, sizeof(cmd), "smbpasswd -L -e %s", user);
                    run_command(cmd);
                    strcpy(users[user_count++], user);
                }
            }
        }
    }
    
    // Build config file
    FILE *fp = fopen("/etc/samba/smb.conf", "w");
    if(fp) {
        fprintf(fp, "[global]\n");
        fprintf(fp, "   workgroup = WORKGROUP\n");
        fprintf(fp, "   server role = standalone server\n");
        fprintf(fp, "   security = user\n");
        fprintf(fp, "   map to guest = bad user\n");
        fprintf(fp, "   smb encrypt = required\n");
        fprintf(fp, "   min protocol = SMB2_02\n");
        fprintf(fp, "   unix password sync = yes\n\n");
        fprintf(fp, "[%s]\n", share_name);
        fprintf(fp, "   path = %s\n", share_path);
        fprintf(fp, "   browseable = %s\n", browseable);
        fprintf(fp, "   writable = %s\n", writable);
        fprintf(fp, "   guest ok = %s\n", guest_ok);
        
        if(user_count > 0 && strcasecmp(guest_ok, "no") == 0) {
            fprintf(fp, "   valid users = ");
            for(int i = 0; i < user_count; i++) {
                fprintf(fp, "%s ", users[i]);
            }
            fprintf(fp, "\n");
        }
        
        fclose(fp);
    }
    
    // Set permissions
    if(user_count > 0 && strcasecmp(guest_ok, "no") == 0) {
        char cmd[MAX_INPUT * 2];
        snprintf(cmd, sizeof(cmd), "chown -R %s %s", users[0], share_path);
        run_command(cmd);
        snprintf(cmd, sizeof(cmd), "chmod 2770 %s", share_path);
        run_command(cmd);
    } else {
        char cmd[MAX_INPUT * 2];
        snprintf(cmd, sizeof(cmd), "chmod 2777 %s", share_path);
        run_command(cmd);
    }
    
    run_command("systemctl enable smbd");
    run_command("systemctl restart smbd");
    
    printf("Setup complete!\n");

    // Retrieve and display SMB hashes
    get_smb_hashes();
    
    return 0;
}
